### PR TITLE
Add composite aggregator

### PR DESCRIPTION
**Pull request by jimczi**
This change adds a module called `aggs-composite` that defines a new aggregation named `composite`.
The `composite` aggregation is a multi-buckets aggregation that creates composite buckets made of multiple sources.
The sources for each bucket can be defined as:
  * A `terms` source, values are extracted from a field or a script.
  * A `date_histogram` source, values are extracted from a date field and rounded to the provided interval.
This aggregation can be used to retrieve **al**l buckets of a deeply nested aggregation by flattening the nested aggregation in composite buckets.
A composite buckets is composed of one value per source and is built for each document as the combinations of values in the provided sources.
For instance the following aggregation:

````
"test_agg": {
  "terms": {
    "field": "field1"
  },
  "aggs": {
    "nested_test_agg":
      "terms": {
        "field": "field2"
      }
  }
}
````
... which retrieves the top N terms for `field1` and for each top term in `field1` the top N terms for `field2`, can be replaced by a `composite` aggregation in order to retrieve **all** the combinations of `field1`, `field2` in the matching documents:

````
"composite_agg": {
  "composite": {
    "sources": [
      {
        "terms": {
          "field": "field1"
          }
      },
      {
        "terms": {
          "field": "field2"
        }
      }
     ]
    }
  }
````

The response of the aggregation looks like this:

````
"aggregations": {
  "composite_agg": {
    "buckets": [
      {
        "key": [
          "alabama",
          "almanach"
        ],
        "doc_count": 100
      },
      {
        "key": [
          "alabama",
          "calendar"
        ],
        "doc_count": 1
      },
      {
        "key": [
          "arizona",
          "calendar"
        ],
        "doc_count": 1
      }
    ]
  }
}
````

By default this aggregation returns 10 buckets sorted in ascending order of the composite key.
Pagination can be achieved by providing `after` values, the values of the composite key to aggregate after.
For instance the following aggregation will aggregate all composite keys that sorts after `arizona, calendar`:

````
"composite_agg": {
  "composite": {
    "after": ["alabama", "calendar"],
    "size": 100,
    "sources": [
      {
        "terms": {
          "field": "field1"
          }
      },
      {
        "terms": {
          "field": "field2"
        }
      }
     ]
    }
  }
````

This aggregation is optimized for indices that set an index sorting that match the composite source definition.
For instance the aggregation above could run faster on indices that defines an index sorting like this:

````
"settings": {
  "index.sort.field": ["field1", "field2"]
}
````

In this case the `composite` aggregation can early terminate on each segment.
This aggregation also accepts multi-valued field but disables early termination for these fields even if index sorting matches the sources definition.
This is mandatory because index sorting picks only one value per document to perform the sort.

For sorted indices, we could jump directly to documents that sort after the provided `after` values in each segment in order to speed up the collection but it can be done in a follow up. 

This aggregation also accepts any sub aggregations and returns the result inside each composite bucket like any other multi-buckets agg:
````
"composite_agg": {
  "composite": {
    "size": 100,
    "sources": [
      {
        "terms": {
          "field": "field1"
          }
      },
      {
        "terms": {
          "field": "field2"
        }
      }
     ]
    },
    "max_value": {
      "max": {
        "field": "number"
     }
  }
}
````

Documentation is missing which is why this is still a WIP.

### CODE REVIEW

File: modules/aggs-composite/src/main/java/org/elasticsearch/search/aggregations/composite/CompositeAggregator.java [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+
+    @Override
+    protected void doPostCollection() throws IOException {
+        if (leaf != null) {
+            leaf.docIdSet = builder.build();
+            contexts.add(leaf);
+        }
+    }
+
+    private LeafBucketCollector getFirstPassCollector() {
```
Comment (martijnvg) [inline modules/aggs-composite/src/main/java/org/elasticsearch/search/aggregations/composite/CompositeAggregator.java]: I think this method and the next method can just return `LeafCollector`? 

The bucket ord isn't used in this implementations. Also these collectors are not directly used by the aggs framework, but are wrapped by a `LeafBucketCollector` instance in `CompositeValuesSource.java`.

File: modules/aggs-composite/src/main/java/org/elasticsearch/search/aggregations/composite/CompositeAggregator.java [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+        grow(keys.size()+1);
+        for (LeafContext context : contexts) {
+            DocIdSetIterator docIdSetIterator = context.docIdSet.iterator();
+            if (docIdSetIterator == null) {
+                continue;
+            }
+            final LeafBucketCollector collector = array.getLeafCollector(context.ctx, getSecondPassCollector(context.subCollector));
+            int docID;
+            while ((docID = docIdSetIterator.nextDoc()) != DocIdSetIterator.NO_MORE_DOCS) {
+                collector.collect(docID, 0L);
```
Comment (martijnvg) [inline modules/aggs-composite/src/main/java/org/elasticsearch/search/aggregations/composite/CompositeAggregator.java]: maybe use `collector.collect(docID);` instead?

File: modules/aggs-composite/src/main/java/org/elasticsearch/search/aggregations/composite/CompositeValuesSource.java [hunk 3]
```diff
@@ -0,0 +1,418 @@
...
+import org.elasticsearch.search.sort.SortOrder;
+
+import java.io.IOException;
+
+import static org.apache.lucene.index.SortedSetDocValues.NO_MORE_ORDS;
+
+/**
+ * A wrapper for {@link ValuesSource} that can record and compare values produced during a collection.
+ */
+abstract class CompositeValuesSource<VS extends ValuesSource, T extends Comparable<T>> {
```
Comment (martijnvg) [inline modules/aggs-composite/src/main/java/org/elasticsearch/search/aggregations/composite/CompositeValuesSource.java]: There is some comparing and ordering done here. I wonder if we can incorporate or extend  Lucene's `PriorityQueue` class here?

File: modules/aggs-composite/src/main/java/org/elasticsearch/search/aggregations/composite/HistogramValuesSource.java [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+
+public class HistogramValuesSource extends ValuesSource.Numeric {
+    private final Numeric vs;
+    private final double interval;
+
+    /**
+     *
+     * @param vs The original values source
+     */
+    public HistogramValuesSource(Numeric vs, double interval) {
```
Comment (martijnvg) [inline modules/aggs-composite/src/main/java/org/elasticsearch/search/aggregations/composite/HistogramValuesSource.java]: constructor can be package protected?

File: modules/aggs-composite/src/main/java/org/elasticsearch/search/aggregations/composite/RoundingValuesSource.java [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+public class RoundingValuesSource extends ValuesSource.Numeric {
+    private final ValuesSource.Numeric vs;
+    private final Rounding rounding;
+
+    /**
+     *
+     * @param vs The original values source
+     * @param rounding How to round the values
+     */
+    public RoundingValuesSource(Numeric vs, Rounding rounding) {
```
Comment (martijnvg) [inline modules/aggs-composite/src/main/java/org/elasticsearch/search/aggregations/composite/RoundingValuesSource.java]: constructor can be package protected?

File: modules/aggs-composite/src/main/java/org/elasticsearch/search/aggregations/composite/CompositeAggregator.java [hunk 6]
```diff
@@ -0,0 +1,221 @@
...
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.TreeMap;
+
+class CompositeAggregator extends BucketsAggregator {
```
Comment (martijnvg) [inline modules/aggs-composite/src/main/java/org/elasticsearch/search/aggregations/composite/CompositeAggregator.java]: make class final?

File: modules/aggs-composite/src/main/java/org/elasticsearch/search/aggregations/composite/CompositeKey.java [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+import java.util.Arrays;
+import java.util.List;
+
+/**
+ * A key that is composed of multiple {@link Comparable} values.
+ */
+public class CompositeKey {
+    final Comparable<?>[] values;
+
+    public CompositeKey(Comparable<?>... values) {
```
Comment (martijnvg) [inline modules/aggs-composite/src/main/java/org/elasticsearch/search/aggregations/composite/CompositeKey.java]: Can the constructor be package protected?

File: modules/aggs-composite/src/main/java/org/elasticsearch/search/aggregations/composite/CompositeValuesSource.java [hunk 8]
```diff
@@ -0,0 +1,418 @@
...
+            return values[slot];
+        }
+
+        @Override
+        LeafBucketCollector getLeafCollector(LeafReaderContext context, LeafBucketCollector next) throws IOException {
+            final SortedBinaryDocValues dvs = vs.bytesValues(context);
+            return new LeafBucketCollector() {
+                @Override
+                public void collect(int doc, long bucket) throws IOException {
+                    if (dvs.advanceExact(doc)) {
```
Comment (martijnvg) [inline modules/aggs-composite/src/main/java/org/elasticsearch/search/aggregations/composite/CompositeValuesSource.java]: The bucket variable is always zero here, right? If that is the case then maybe we should have assertions for this in these LeafBucketCollector anonymous classes?

File: modules/aggs-composite/src/main/java/org/elasticsearch/search/aggregations/composite/InternalComposite.java [hunk 9]
```diff
@@ -0,0 +1,282 @@
...
+import java.util.PriorityQueue;
+
+public class InternalComposite extends InternalMultiBucketAggregation<InternalComposite, InternalComposite.InternalBucket>
+            implements CompositeAggregation {
+
+    private final int size;
+    private final List<InternalBucket> buckets;
+    private final int[] reverseMuls;
+
+    public InternalComposite(String name, int size, List<InternalBucket> buckets, int[] reverseMuls,
```
Comment (martijnvg) [inline modules/aggs-composite/src/main/java/org/elasticsearch/search/aggregations/composite/InternalComposite.java]: these constructors can be package protected?

File: modules/aggs-composite/src/test/java/org/elasticsearch/search/aggregations/composite/CompositeAggregationBuilderTests.java [hunk 10]
```diff
@@ -0,0 +1,181 @@
...
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+
+import static org.elasticsearch.test.EqualsHashCodeTestUtils.checkEqualsAndHashCode;
+import static org.hamcrest.Matchers.hasSize;
+
+public class CompositeAggregationBuilderTests extends ESTestCase {
```
Comment (martijnvg) [inline modules/aggs-composite/src/test/java/org/elasticsearch/search/aggregations/composite/CompositeAggregationBuilderTests.java]: Maybe add a dedicated testcase for each build (extending from `AbstractStreamableTestCase`)?

**Review by martijnvg — COMMENTED** [review]
In general this change looks good to me. I left some questions and remarks. I'll do another review round later.

This PR is quite large and it would be great if someone else can also take a look at it. Maybe @colings86?

File: modules/aggs-composite/src/main/java/org/elasticsearch/search/aggregations/composite/CompositeKey.java (hunk 7 above)
Comment (jimczi) [reply modules/aggs-composite/src/main/java/org/elasticsearch/search/aggregations/composite/CompositeKey.java]: It can be used to build a composite aggregation so I think it's useful to keep it public for now. It adds the ability to start from anywhere a composite aggregation in the java client.

File: modules/aggs-composite/src/main/java/org/elasticsearch/search/aggregations/composite/CompositeValuesSource.java (hunk 3 above)
Comment (jimczi) [reply modules/aggs-composite/src/main/java/org/elasticsearch/search/aggregations/composite/CompositeValuesSource.java]: I am using this class to create slots that can be referenced in a priority queue. It's mainly a comparator + a composite array that can update the values. It is equivalent to a `FieldComparator` but for buckets.

File: modules/aggs-composite/src/test/java/org/elasticsearch/search/aggregations/composite/CompositeAggregationBuilderTests.java (hunk 10 above)
Comment (jimczi) [reply modules/aggs-composite/src/test/java/org/elasticsearch/search/aggregations/composite/CompositeAggregationBuilderTests.java]: This is a test for the `CompositeAggregationBuilder` so it randomizes the different value sources that can be used inside. Splitting the test for each source would defeat the purpose since the idea is to compose an aggregation from different value source.

File: modules/aggs-composite/src/main/java/org/elasticsearch/search/aggregations/composite/CompositeAggregator.java [hunk 11]
```diff
@@ -0,0 +1,224 @@
...
+        builder = new RoaringDocIdSet.Builder(ctx.reader().maxDoc());
+        final CompositeValuesSource.Collector inner = array.getLeafCollector(ctx, getFirstPassCollector());
+        return new LeafBucketCollector() {
+            @Override
+            public void collect(int doc, long zeroBucket) throws IOException {
+                assert zeroBucket == 0L;
+                inner.collect(doc);
+            }
+        };
+    }
```
Comment (colings86) [inline modules/aggs-composite/src/main/java/org/elasticsearch/search/aggregations/composite/CompositeAggregator.java]: I think its worth adding some JAvaDocs here to explain how the collection works and what is happening in the first pass and the second pass

File: modules/aggs-composite/src/main/java/org/elasticsearch/search/aggregations/composite/CompositeValuesSourceBuilder.java [hunk 12]
```diff
@@ -0,0 +1,288 @@
...
+import org.elasticsearch.search.internal.SearchContext;
+import org.elasticsearch.search.sort.SortOrder;
+
+import java.io.IOException;
+import java.util.Objects;
+
+/**
+ * A {@link ValuesSource} builder for {@link CompositeAggregationBuilder}
+ */
+public abstract class CompositeValuesSourceBuilder<AB extends CompositeValuesSourceBuilder<AB>> implements Writeable, ToXContentFragment {
```
Comment (colings86) [inline modules/aggs-composite/src/main/java/org/elasticsearch/search/aggregations/composite/CompositeValuesSourceBuilder.java]: This looks very like the ValuesSourceBuilder that already exists. Are we not able to use that here?

**Review by colings86 — COMMENTED** [review]
I took a first look, will need to review again later.

One thing that I am a little wary of is that we seem to be recreating a lot of the ValuesSource classes here. I am wondering if we can reuse more of the existing ValuesSource builder and parser code to avoid having it in two places? We will want the `sources` to feel very like an array of ValuesSource configs so as far as possible it would be good to use the same code to parse them as we have for the regular aggs.

File: modules/aggs-composite/src/main/java/org/elasticsearch/search/aggregations/composite/CompositeValuesSourceBuilder.java (hunk 12 above)
Comment (jimczi) [reply modules/aggs-composite/src/main/java/org/elasticsearch/search/aggregations/composite/CompositeValuesSourceBuilder.java]: I tried but this values source builder is not an `AbstractAggregationBuilder`. It is used solely to access the values, not to build a complete aggregator. I also need to access the values in a specific way in order to be able to build the combinations of multiple values source so my usage is really an edge case.

File: modules/aggs-composite/src/main/java/org/elasticsearch/search/aggregations/composite/CompositeAggregator.java (hunk 11 above)
Comment (jimczi) [reply modules/aggs-composite/src/main/java/org/elasticsearch/search/aggregations/composite/CompositeAggregator.java]: ++

**Comment by jimczi** [comment]
@martijnvg @colings86 thanks for reviewing
I pushed more commits to address your comments.
I also pushed the documentation for this aggregation, @clintongormley can you take a look and validate the API and response ?

File: docs/reference/aggregations/bucket/composite-aggregation.asciidoc [hunk 13]
```diff
@@ -0,0 +1,557 @@
+[[search-aggregations-bucket-composite-aggregation]]
+=== Composite Aggregation
+
+experimental[]
+
+A multi-bucket aggregation that creates composite buckets from different sources.
+The buckets are build from the combinations of the values extracted/created for each document and each
```
Comment (clintongormley) [inline docs/reference/aggregations/bucket/composite-aggregation.asciidoc]: build -> built

File: docs/reference/aggregations/bucket/composite-aggregation.asciidoc [hunk 14]
```diff
@@ -0,0 +1,557 @@
...
+                }
+            }
+        }
+    }
+}
+--------------------------------------------------
+// CONSOLE
+
+WARNING: The optimization takes effect only if the fields used for sorting are single-valued and follow
+the same order than the aggregation (`desc` or `asc`).
```
Comment (clintongormley) [inline docs/reference/aggregations/bucket/composite-aggregation.asciidoc]: than -> as

**Comment by clintongormley** [comment]
This is looking good.  I wonder if `sources` should be named, eg:

```
    "sources": [
        { "date": {"date_histogram": { "field": "timestamp", "interval": "1d" }}},
        { "product": {"terms": {"field": "product"}}}
    ]
```

Then the result would look like this:

```
{
     ...
     "aggregations": {
         "my_buckets": {
             "buckets": [
                 {
                     "values": {
                         "date": 1494201600000,
                         "product": rocky"
                     },
                     "doc_count": 1
                 },
                 {    
                     "values": {
                         "date": 1494288000000,
                         "product": mad max"
                     },
                     "doc_count": 2
                 }
             ]
         }
     }
 }
```

This seems more human readable, but maybe using a hashmap instead of an array to represent the values is less useful for consumers of this API like Kibana.  I'm easy.

The only other thing I'd suggest is to add an introductory paragraph at the beginning of the docs explaining why you would use the `composite` agg, ie extolling its benefits.  They're all mentioned on the page, but the user has to read the whole page before they realise how powerful this agg is.

File: modules/aggs-composite/src/main/java/org/elasticsearch/search/aggregations/composite/CompositeValuesSourceConfig.java [hunk 15]
```diff
@@ -0,0 +1,47 @@
...
+        this.vs = vs;
+        this.canEarlyTerminate = canEarlyTerminate;
+        this.reverseMul = order == SortOrder.ASC ? 1 : -1;
+    }
+
+    ValuesSource valuesSource() {
+        return vs;
+    }
+
+    int reverseMul() {
```
Comment (colings86) [inline modules/aggs-composite/src/main/java/org/elasticsearch/search/aggregations/composite/CompositeValuesSourceConfig.java]: Could you add a javaDoc for this. It took me a while to work out what this actually was as it wasn't obvious from where its used

File: modules/aggs-composite/src/main/java/org/elasticsearch/search/aggregations/composite/DateHistogramValuesSourceBuilder.java [hunk 16]
```diff
@@ -0,0 +1,200 @@
...
+                // the interval is a time value?
+                tzRoundingBuilder = Rounding.builder(
+                    TimeValue.parseTimeValue(dateHistogramInterval.toString(), null, getClass().getSimpleName() + ".interval"));
+            }
+        } else {
+            // the interval is an integer time value in millis?
+            tzRoundingBuilder = Rounding.builder(TimeValue.timeValueMillis(interval));
+        }
+        Rounding rounding = tzRoundingBuilder.build();
+        return rounding;
```
Comment (colings86) [inline modules/aggs-composite/src/main/java/org/elasticsearch/search/aggregations/composite/DateHistogramValuesSourceBuilder.java]: Should we have timezone support here?

**Review by colings86 — COMMENTED** [review]
@jimczi I left a few more comments

**Comment by polyfractal** [comment]
++ to some kind of ability to name sources.  It'll be a lot easier for applications to use if they can refer to names rather than positions.  Ditto to the `after` parameter, maybe a map instead of ordered array?

```
"after": {
  "date": 1494201600000, 
  "product": "rocky"
}
```

Two questions! :)

If a bucket doesn't have any documents, is it returned in results with `doc_count: 0`, or does that composite key not show up at all?

My understanding is that if index sorting is not enabled, paging through the results with `after` means it has to re-evaluate all the docs again (since it can't "jump" to the right position).  Do you happen to know how expensive that actually is?  I'm thinking about the case where you have a very large index and need to deeply page into it.  Maybe the cost isn't so bad because doc value evaluation is fast?

**Comment by jimczi** [comment]
> If a bucket doesn't have any documents, is it returned in results with doc_count: 0, or does that composite key not show up at all?

The composite buckets are created from existing values only. 
This aggregation implicitly sets `min_doc_count` to 1 and this value cannot be changed. 
I guess that this is more an issue for `histogram` and `date_histogram` sources. Do you think it is worth trying to add the support to create empty buckets for these sources ?

> My understanding is that if index sorting is not enabled, paging through the results with after means it has to re-evaluate all the docs again (since it can't "jump" to the right position). Do you happen to know how expensive that actually is? I'm thinking about the case where you have a very large index and need to deeply page into it. Maybe the cost isn't so bad because doc value evaluation is fast?

Your understanding is correct. We need to reevaluate all documents on every requests but the memory consumption (dictated by `size`) remains the same no matter how far you are in the buckets. I think it's ok because this aggregation should be used in a batch context, it's not a realtime thing so I consider this as a safe way to dump all buckets from a complex aggregation tree without killing your cluster. We have some ideas to make it faster when the index is not sorted (related to https://github.com/elastic/elasticsearch/issues/23022) but for  the moment the order of magnitude between a sorted index and a non sorted index can be huge. 
It also depends on the number of buckets that you can handle per query, in a sense this is the same as scrolling an index with a specific sort that do not depends on doc_id.

**Comment by polyfractal** [comment]
Thanks @jimczi!

> Do you think it is worth trying to add the support to create empty buckets for these sources ?

I'm not sure... probably not.  I was mostly just curious, not sure it's really needed.  The main reason we went to `min_doc_count: 0` was for pipeline aggs which isn't really a consideration here.

 > Your understanding is correct. We need to reevaluate all documents on every requests but the memory consumption (dictated by size) remains the same no matter how far you are in the buckets.

Good to know, thanks for the explanation.  I'll run some tests here locally to get a feel for the impact.  :)

**Comment by clintongormley** [comment]
> The main reason we went to min_doc_count: 0 was for pipeline aggs which isn't really a consideration here.

Does this mean that pipeline aggs are not supported under the `composite` agg?

**Comment by colings86** [comment]
@clintongormley it'll mean that pipeline aggregations like derivative, moving_average and cumulative sum which only work with histogram or date_histogram aggregations will not work but we wouldn't be able to make them work across pages anyway since it would need to maintain state across requests. Pipeline aggregations like bucket_selector and bucket_script I think will still work.

File: core/src/main/java/org/elasticsearch/index/mapper/CompletionFieldMapper.java [hunk 17]
```diff
@@ -546,7 +547,8 @@ private void parse(ParseContext parseContext, Token token, XContentParser parser
                 }
             }
         } else {
-            throw new ElasticsearchParseException("failed to parse expected text or object got" + token.name());
+            throw new ParsingException(parser.getTokenLocation(), "failed to parse expected text or object got " + token.name());
```
Comment (martijnvg) [inline core/src/main/java/org/elasticsearch/index/mapper/CompletionFieldMapper.java]: This change looks unrelated? and I think was addressed in a different PR.

**Comment by jimczi** [comment]
I pushed another iteration to address reviews.
The main changes are:

- The composite key is now rendered as a map (thanks @clintongormley !)
- The `date_histogram` handles time_zone
- Updated documentation

@martijnvg can you take another look ?

File: modules/aggs-composite/src/main/java/org/elasticsearch/search/aggregations/composite/CompositeKey.java [hunk 18]
```diff
@@ -0,0 +1,84 @@
...
+        CompositeKey that = (CompositeKey) o;
+        return Arrays.equals(values, that.values);
+    }
+
+    @Override
+    public int hashCode() {
+        return Arrays.hashCode(values);
+    }
+
+    static String formatValue(Object value, DocValueFormat formatter) {
```
Comment (martijnvg) [inline modules/aggs-composite/src/main/java/org/elasticsearch/search/aggregations/composite/CompositeKey.java]: removed? It does not seem to be used.

File: core/src/main/java/org/elasticsearch/index/mapper/DateFieldMapper.java [hunk 19]
```diff
@@ -182,7 +182,7 @@ public TypeParser() {
         protected FormatDateTimeFormatter dateTimeFormatter;
         protected DateMathParser dateMathParser;
 
-        DateFieldType() {
+        public DateFieldType() {
```
Comment (martijnvg) [inline core/src/main/java/org/elasticsearch/index/mapper/DateFieldMapper.java]: I prefer to keep this package protected. I think in `CompositeAggregatorTests#setUp()` we should do this instead:

```java
DateFieldMapper.Builder builder = new DateFieldMapper.Builder("date");
builder.docValues(true);
DateFieldMapper fieldMapper = builder.build(new Mapper.BuilderContext(createIndexSettings().getSettings(), new ContentPath(0)));
FIELD_TYPES[3] = fieldMapper.fieldType();
```

**Review by martijnvg — APPROVED** [review]
Left two small comments. LGTM otherwise!
